include/veriblock/blockchain/pop/pop_state_machine.hpp

```diff
@@ -96,7 +122,7 @@ struct PopStateMachine {
 
     for (auto* index : chain) {
       if (!index->isValid() || !applyBlock(*index, state)) {
-        unapply(*index, from);
+        unapply(*index->pprev, from);
```

Comment:
you can not do that, because index is not guaranteed to be in the middle of chain, it can be very first block

Reply:
also chain may have size=1

Reply:
the chain starts at from.height + 1, thus index->pprev of the first block in the chain == from; unapply(from, from) is a no-op.